src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -100,18 +100,18 @@ public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
             // Use local pipe to avoid blocking the current thread on networking IO.
             var localPipe = new Pipe(PipeOptionsWithUnlimitedWriterBuffer);
 
-            Exception? exception = null;
-
             // start a task on a thread pool thread copying from the RPC pipe to a local pipe:
             var copyTask = Task.Run(async () =>
             {
+                Exception? exception = null;
+
                 try
                 {
                     await pipeReader.CopyToAsync(localPipe.Writer, cancellationToken).ConfigureAwait(false);
                 }
-                catch (Exception e)
+                catch (Exception e) when ((exception = e) == null) // capture, but don't catch the exception
```

Comment:
i don't get it... what does capturing accomplish here.  it looks like you write to the local.. but nothing ever reads from it.

Reply:
It's read in the finally block.

Reply:
ok.  i see that now (wasn't showing in the diff since that read already exists).  However, i think we need a *lot* more comments here.  Your comment is accurate.  We capture... but don't catch the exception.  But taht simply explains what is happening, and not at all 'why'.  I say this honestly, but i have no idea why it's not right to propagate teh exception.  If we're reading... and it fails with an exception... why not propagate it?  

I view this code as being as critical as AsyncLazy.  In other words, it's incredibly subtle, but utterly paramount to get right.  Because of htat subtlety, i think these pieces need a *huge* amount of explanation as to what's going on.

Thanks!

Reply:
We do propagate the exception (that's the point of the change).